contrib_bots/lib/gcalendar.py

```diff
@@ -0,0 +1,356 @@
+# Zulip's roadmap.
+
+def get_credentials():
+    '''
+    Gets valid user credentials from storage.
+
+    If nothing has been stored, or if the stored credentials are invalid,
+    the OAuth2 flow is completed to obtain the new credentials.
+
+    Returns:
+        Credentials, the obtained credential.
+    '''
+    home_dir = os.path.expanduser('~')
+    credential_dir = os.path.join(home_dir, '.gcalbot-credentials')
+    if not os.path.exists(credential_dir):
+        os.makedirs(credential_dir)
+    credential_path = os.path.join(credential_dir, 'gcalbot.json')
+
+    store = Storage(credential_path)
+    cred = store.get()
+    if not cred or cred.invalid:
+        flow = client.flow_from_clientsecrets(os.path.join(home_dir, CLIENT_SECRET_FILE), SCOPES)
+        flow.user_agent = APPLICATION_NAME
+        flags = argparser.parse_args('--noauth_local_webserver --logging_level INFO'.split())
+        cred = tools.run_flow(flow, store, flags)
+    return cred
+
+def reply_message(msg, original_msg, client):
+    '''
+    Replies to a specific message, through the channel of origin.
+
+    Keyword arguments:
+    msg -- content of the reply
+    original_msg -- message that the bot replies to
+    client -- client from the message handler
+    '''
+    message_type = original_msg['type']
+
+    message_data = {'content': msg, 'type': message_type}
+
+    if message_type == 'private':
+        message_data.update({'to': original_msg['sender_email']})
+    elif message_type == 'stream':
+        message_data.update({'to': original_msg['display_recipient'],
+                             'subject': original_msg['subject']})
+    else:
+        raise UnknownMessageTypeError(message_type)
+
+    client.send_message(message_data)
+
+def parse_message(message):
+    '''
+    Identifies and parses the different parts of a message sent to the bot
+
+    Returns:
+        Values, a tuple that contains the 2 (or 3) parameters of the message
+    '''
+    content = message.replace('@gcalendar ', '')
+
+    try:
+        splits = content.split('|')
+        title = splits[0].strip()
+
+        settings = {
+            'RETURN_AS_TIMEZONE_AWARE': True
+        }
+
+        if len(splits) == 4: # Delimited event with timezone
+            settings['TIMEZONE'] = splits[3].strip()
+
+        start_date = dateparser.parse(splits[1].strip(), settings=settings)
+        end_date = None
+
+        if len(splits) >= 3: # Delimited event
+            end_date = dateparser.parse(splits[2].strip(), settings=settings)
+
+        if not start_date or (len(splits) >= 3 and not end_date):  # Unknown date format
+            raise MessageParseError(2)
+
+        # Notice there isn't a "full day event with timezone", because the
+        # timezone is irrelevant in that type of events
+    except IndexError as e:
+        logging.exception(e)
+        return MessageParseError(1), None, None
+    except MessageParseError as e:
```

Comment:
You can simplify your code if you remove this method and catch `MessageParseError` in `handle_message`.